pkg/plugin/plugin.go

```diff
@@ -0,0 +1,113 @@
+package plugin
+
+import (
+	"fmt"
+	"os"
+	"os/exec"
+	"runtime"
+	"strings"
+	"syscall"
+)
+
+// Handler is capable of parsing command line arguments
+// and performing executable filename lookups to search
+// for valid plugin files, and execute found plugins.
+type Handler interface {
+	// exists at the given filename, or a boolean false.
+	// Lookup will iterate over a list of given prefixes
+	// in order to recognize valid plugin filenames.
+	// The first filepath to match a prefix is returned.
+	Lookup(filename string) (string, bool)
+	// Execute receives an executable's filepath, a slice
+	// of arguments, and a slice of environment variables
+	// to relay to the executable.
+	Execute(executablePath string, cmdArgs, environment []string) error
+}
+
+// DefaultPluginHandler implements PluginHandler
+type DefaultPluginHandler struct {
+	ValidPrefixes []string
+}
+
+// NewDefaultPluginHandler instantiates the DefaultPluginHandler with a list of
+// given filename prefixes used to identify valid plugin filenames.
+func NewDefaultPluginHandler(validPrefixes []string) *DefaultPluginHandler {
+	return &DefaultPluginHandler{
+		ValidPrefixes: validPrefixes,
+	}
+}
+
+// Lookup implements PluginHandler
+func (h *DefaultPluginHandler) Lookup(filename string) (string, bool) {
+	for _, prefix := range h.ValidPrefixes {
+		path, err := exec.LookPath(fmt.Sprintf("%s-%s", prefix, filename))
+		if err != nil || len(path) == 0 {
+			continue
+		}
+		return path, true
+	}
+
+	return "", false
+}
+
+// Execute implements PluginHandler
+func (h *DefaultPluginHandler) Execute(executablePath string, cmdArgs, environment []string) error {
+
+	// Windows does not support exec syscall.
+	if runtime.GOOS == "windows" {
+		// #nosec G204
+		cmd := exec.Command(executablePath, cmdArgs...)
+		cmd.Stdout = os.Stdout
+		cmd.Stderr = os.Stderr
+		cmd.Stdin = os.Stdin
+		cmd.Env = environment
+		err := cmd.Run()
+		if err == nil {
+			os.Exit(0)
```

Comment:
I find it a bit confusing that this function _sometimes_ returns an error to the caller and _sometimes_ doesn't return (via exec syscall or Exit here).

- One option is to make the contract "this function doesn't return", and do os.Exit() on all errors.
    - In particular, [cmd.Run](https://golang.org/pkg/os/exec/#Cmd.Run) may return `ExitError` when plugin was launched successfully but exited with non-0 status, say 42.  
      In the UNIX `Exec` code path, the parent of `ocm` gets all exit statuses directly from the process that replaced it.  As currently written, on Windows, this function returns an error and main.go does Exit(1), losing the plugin's exact status.  
      So for this option, you should handle `ExitError` here by exiting with same status.

- Another clean option is to decide this function always returns.
    - This means you can't use exec syscall, always exec.Command that waits for subprocess.  
      IMHO that's good idea anyway for code simplicity and testing :+1:.
    - main.go then should detect ExitError and propogate `ExitError.ExitCode()`.

  P.S. `sycall` package is [frozen & deprecated](https://golang.org/s/go1.4-syscall).  If you want it, [x/sys/unix.Exec](https://godoc.org/golang.org/x/sys/unix#Exec) is preferred.  (though I expect sycall.Exec to work forever...)